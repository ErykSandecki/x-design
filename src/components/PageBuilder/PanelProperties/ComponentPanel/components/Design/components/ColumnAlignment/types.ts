// types
import { AlignmentHorizontal, AlignmentVertical } from 'types';
import { TButtonGroup } from 'shared/UITools/ButtonGroup/types';

export type TAligmentHorizontalButtonGroup = Array<
  Pick<TButtonGroup, 'name' | 'tooltip'> & { key: AlignmentHorizontal }
>;
export type TAligmentVerticalButtonGoup = Array<Pick<TButtonGroup, 'name' | 'tooltip'> & { key: AlignmentVertical }>;
