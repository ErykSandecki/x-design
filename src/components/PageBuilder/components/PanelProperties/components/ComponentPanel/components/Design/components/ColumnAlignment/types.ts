// types
import { AlignmentHorizontal, AlignmentVertical } from 'types';
import { TButtonGroup } from 'shared/UITools/components/ButtonGroup/types';

export type TAligmentHorizontalButtonGroup = Array<Pick<TButtonGroup, 'name'> & { key: AlignmentHorizontal }>;
export type TAligmentVerticalButtonGoup = Array<Pick<TButtonGroup, 'name'> & { key: AlignmentVertical }>;
