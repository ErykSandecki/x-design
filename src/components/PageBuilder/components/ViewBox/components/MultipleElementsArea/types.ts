// types
import { TElement } from 'types';

export type TCoordinatesData = {
  elementsCoordinates: Array<{
    coordinates: TRectCoordinates;
    id: TElement['id'];
  }>;
  outline: TRectCoordinates;
};
