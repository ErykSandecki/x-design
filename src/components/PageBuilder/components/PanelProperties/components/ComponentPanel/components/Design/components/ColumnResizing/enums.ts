export enum PopoverItem {
  auto = 1,
  fixed = 0,
  unit = 2,
}
