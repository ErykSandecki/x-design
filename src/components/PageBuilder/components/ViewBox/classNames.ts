// utils
import { composeClassNames } from 'utils';

export const className = 'ViewBox';

export const classNames = composeClassNames(className, [className, 'createFrame'] as const);
