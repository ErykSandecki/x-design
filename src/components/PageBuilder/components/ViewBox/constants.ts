// others
import { translationNameSpace as parentNameSpace } from '../../constants';

export const THROTTLE_WAIT = 20;
export const translationNameSpace = `${parentNameSpace}.viewBox`;
