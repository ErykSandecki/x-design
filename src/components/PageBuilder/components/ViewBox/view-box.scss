@import '../../styles/z-indexes';

.ViewBox {
  &--create-frame {
    cursor: crosshair;
  }
}
