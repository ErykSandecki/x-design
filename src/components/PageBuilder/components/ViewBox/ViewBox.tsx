import { FC, memo } from 'react';
import { useDispatch, useSelector } from 'react-redux';

// components
import Elements from './components/Elements/Elements';
import ElementArea from './components/ElementArea/ElementArea';
import MultipleElementsArea from './components/MultipleElementsArea/MultipleElementsArea';
import OverlayContainer from './components/OverlayContainer/OverlayContainer';
import SelectableArea from './components/SelectableArea/SelectableArea';
import { ZoomBox } from 'shared';

// core
import { useRefs } from 'pages/PageBuilderPage/core/RefsProvider';

// others
import { className, classNames } from './classNames';

// hooks
import { useTheme } from 'hooks';
import { useViewBoxEvents } from './hooks/useViewBoxEvents';

// store
import { eventSelectorCreator, pageBackgroundSelectorCreator } from 'store/pageBuilder/selectors';
import { setAreCoordinates } from 'store/pageBuilder/actions';

// styles
import styles from './view-box.scss';

// types
import { MouseMode } from 'types/enums/mouseMode';
import { TColor } from 'types';

export type TViewBoxProps = {
  coordinates: T3DCoordinates;
  mouseMode: MouseMode;
  setCoordinates: (coordinates: T3DCoordinates) => void;
  setMouseMode: (mouseMode: MouseMode) => void;
};

const ViewBox: FC<TViewBoxProps> = ({ coordinates, mouseMode, setCoordinates, setMouseMode }) => {
  const colorSampler = useSelector(eventSelectorCreator('colorSampler'));
  const background = useSelector(pageBackgroundSelectorCreator('-1'));
  const data = background.properties as TColor;
  const dispatch = useDispatch();
  const { zoomBoxRef, zoomContentRef } = useRefs();
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);
  const { elementArea, selectableArea, ...events } = useViewBoxEvents(coordinates, mouseMode, setMouseMode);

  return (
    <ZoomBox
      alpha={data.alpha}
      backgroundColor={data.color}
      backgroundVissible={background.visible}
      classes={{
        className: cx(classNamesWithTheme[className].name, [
          classNamesWithTheme[className].modificators.createFrame,
          mouseMode === MouseMode.toolBeltA,
        ]),
      }}
      colorSampler={colorSampler as boolean}
      coordinates={coordinates}
      mouseMode={mouseMode}
      onMouseMoveDepedencies={[elementArea, mouseMode]}
      onMouseUpDepedencies={[elementArea, mouseMode]}
      onUpdateCoordinates={(coordinates) => dispatch(setAreCoordinates(coordinates))}
      setCoordinates={setCoordinates}
      zoomBoxRef={zoomBoxRef}
      zoomContentRef={zoomContentRef}
      {...events}
    >
      <MultipleElementsArea />
      <Elements
        eventsDisabled={elementArea !== null || mouseMode !== MouseMode.default}
        mouseMode={mouseMode}
        parentId="-1"
      />
      <ElementArea elementArea={elementArea} />
      <SelectableArea selectableArea={selectableArea} />
      <OverlayContainer />
    </ZoomBox>
  );
};

export default memo(ViewBox);
