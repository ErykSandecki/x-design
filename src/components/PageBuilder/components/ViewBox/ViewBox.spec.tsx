import { fireEvent, render } from '@testing-library/react';
import { Provider } from 'react-redux';

// components
import ViewBox from './ViewBox';

// core
import { RefsProvider } from 'pages/PageBuilderPage/core/RefsProvider';

// mocks
import { elementAllDataMock } from 'test/mocks/reducer/pageBuilderMock';
import 'test/mocks/sagas/allSagas';

// others
import { BASE_3D } from 'shared';

// store
import { configureStore } from 'store/store';

// types
import { MouseMode } from 'types/enums/mouseMode';
import { E2EAttribute, MouseButton } from 'types';

// utils
import { getByE2EAttribute } from 'test';

const element = document.createElement('div');
const overlayContainer = document.createElement('div');
const mockCallBack = jest.fn();
const sharedRefs = {
  [elementAllDataMock.id]: element,
  ['2']: element,
};

jest.mock('lodash', () => ({
  ...jest.requireActual('lodash'),
  debounce:
    (callback: any) =>
    (value: any): any =>
      callback(value),
}));

describe('ViewBox snapshots', () => {
  beforeAll(() => {
    element.style.height = '100px';
    element.style.width = '100px';
    document.body.appendChild(overlayContainer);
  });

  it('should render ViewBox', () => {
    // mock
    const store = configureStore();

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <RefsProvider
          itemsRefs={sharedRefs}
          overlayContainerRefHtml={overlayContainer}
        >
          <ViewBox
            coordinates={BASE_3D}
            mouseMode={MouseMode.default}
            setCoordinates={mockCallBack}
            setMouseMode={mockCallBack}
          />
        </RefsProvider>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('ViewBox behaviors', () => {
  it('should triiger on update coordinates', () => {
    // mock
    const store = configureStore();

    // before
    const { container } = render(
      <Provider store={store}>
        <RefsProvider
          itemsRefs={sharedRefs}
          overlayContainerRefHtml={overlayContainer}
        >
          <ViewBox
            coordinates={BASE_3D}
            mouseMode={MouseMode.default}
            setCoordinates={mockCallBack}
            setMouseMode={mockCallBack}
          />
        </RefsProvider>
      </Provider>,
    );

    // action
    fireEvent.mouseDown(
      getByE2EAttribute(container, E2EAttribute.box, 'zoom-box'),
      {
        buttons: MouseButton.rmb,
      },
    );
    fireEvent.mouseMove(window, {
      buttons: MouseButton.rmb,
    });

    // result
    expect(mockCallBack.mock.calls.length).toBe(1);
  });
});
