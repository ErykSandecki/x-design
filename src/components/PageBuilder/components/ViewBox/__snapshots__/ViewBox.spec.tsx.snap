// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewBox snapshots should render ViewBox 1`] = `
<DocumentFragment>
  <div
    class="css-mu34gm ViewBox ViewBox--light ZoomBox ZoomBox--light ZoomBox--idle ZoomBox--idle--light ZoomBox--default ZoomBox--default--light"
    data-test-box="zoom-box"
    tabindex="0"
  >
    <div
      class="css-wcnq0t ZoomBox__background-mask ZoomBox__background-mask--light"
      data-test-box=""
      style="background-color: rgb(241, 242, 246); display: initial;"
    />
    <div
      class="css-wcnq0t ZoomBox__texture-blank ZoomBox__texture-blank--light"
      data-test-box=""
      style="transform: translate(0px, 0px) scale(1);"
    />
    <div
      class="css-fw1hyy ZoomBox__zoom-content ZoomBox__zoom-content--light"
      data-test-box=""
      style="transform: translate(0px, 0px) scale(1);"
    >
      <div
        class="css-13fj95k"
        data-test-box=""
        id="overlayContainer"
        style="z-index: var(--xd-page-builder-zindex-overlay-container);"
      />
    </div>
  </div>
</DocumentFragment>
`;
